### ISSUE TITLE
[datagrid] Toolbar panels don't restore focus after closing with `Escape`

### ISSUE THREAD

**Issue by Isaac1606** [mui/mui-x#10241]
### Steps to reproduce 🕹

* For  Density selector: 
   Link to live example: https://mui.com/x/react-data-grid/accessibility/#density
  
  Steps:
  1. Go to the live example link.
  2. Navigate with keyboard or click with the mouse the density button in the DataGrid.
  3. Select any option in the menu or press esc.

* For  GridActionsCellItem with showInMenu enabled: 
    Link to live example: https://mui.com/x/react-data-grid/accessibility/#density
  
  Steps:
  1. Go to the live example link.
  2. Reach the DataGrid example.
  3. Go to the last column.
  4. Press/select the icon with the three dots.
  5. Select any option in the menu or press esc.

### Current behavior 😯

The menu with the set of options disappears (as expected) after one of the options is selected or if the esc button is pressed, but the cursor ceases to exist and there is no focused element whatsoever. 

### Expected behavior 🤔

Focus should return to the button that caused the menu to appear.

### Context 🔦

Give users the capability to navigate throughout the entire DataGrid and options using only Keyboard.

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Used browser: Google Chrome.
   System:
    OS: macOS 13.4
  Binaries:
    Node: 18.17.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 9.8.1 - /usr/local/bin/npm
  Browsers:
    Chrome: 114.0.5735.106
    Edge: Not Found
    Safari: 16.5
  npmPackages:
    @emotion/react: ^11.11.1 => 11.11.1 
    @emotion/styled: ^11.11.0 => 11.11.0 
    @mui/base:  5.0.0-beta.14 
    @mui/core-downloads-tracker:  5.14.8 
    @mui/icons-material: ^5.14.1 => 5.14.8 
    @mui/material: ^5.14.2 => 5.14.8 
    @mui/private-theming:  5.14.8 
    @mui/styled-engine:  5.14.8 
    @mui/system:  5.14.8 
    @mui/types:  7.2.4 
    @mui/utils:  5.14.8 
    @mui/x-data-grid: ^6.10.2 => 6.12.1 
    @types/react: 18.2.17 => 18.2.17 
    react: 18.2.0 => 18.2.0 
    react-dom: 18.2.0 => 18.2.0 
    typescript: 5.1.6 => 5.1.6 
```
</details>

**Comment by MBilalShafi** [comment]
Thanks for raising the issue.

Seems like a similar problem exists with all the menus that open up by clicking on the Toolbar menu buttons and actions menu. Honestly I am not sure what should be the behavior when clicking on the button and pressing escape key or closing the menu.

Focus shifting to the button could certainly be one direction to proceed to but that way the user will still not be able to maneuver through the cells by pressing arrow keys as that happens only when a Grid cell is focused. So pressing tab and moving the focus to the Grid cells will be required before the keyboard navigation starts working with the arrow keys.

To me, it would make more sense in conjunction with https://github.com/mui/mui-x/issues/4658 so the button will also be tabbable and the navigation will feel more natural.

@mui/xgrid What do you think?

**Comment by michelengelen** [comment]
Hey @MBilalShafi 

Considering this section:

> Focus shifting to the button could certainly be one direction to proceed to but that way the user will still not be able to maneuver through the cells by pressing arrow keys as that happens only when a Grid cell is focused. So pressing tab and moving the focus to the Grid cells will be required before the keyboard navigation starts working with the arrow keys.

There is actually a pretty straight forward definition in the W3C standards to be found [here](https://www.w3.org/WAI/ARIA/apg/patterns/menu-button/examples/menu-button-actions/), so the proposed solution from @Isaac1606 is correct. If the menu gets closed (and is not navigating away from the page through an action) the `button` should receive `focus` back from the menu.

I'd like some more opinions on it though.

**Comment by romgrk** [comment]
Yes this is an accessibility violation, focus management is part of the ARIA spec.

**Comment by michelengelen** [comment]
@romgrk if the bot closes this again we should probably move this? Don't know why that got reopened 🤷🏼

**Comment by DanailH** [comment]
@michelengelen we forgot to remove the `status: waiting for follow-up` label when the issue was reopened, which brings me to - there is an opportunity for improvement of the `no-response.yml` action to also remove the `status: waiting for follow-up` label when it closes the issue.

**Comment by romgrk** [comment]
No I think the label should stay, to mark that it's the author's turn to respond. Anyway it's a rare case when we keep it open even if the author loses interest.

**Comment by Isaac1606** [comment]
Hey guys any updates on this? or a temporary workaround? Like a wrote in the context section I need to give the users the capability to navigate throughout the entire DataGrid and options using Keyboard only. This a blocker for me sine I can not release the project without this feature. I would truly appreciate any updates, thanks in advance.

**Comment by romgrk** [comment]
I'll take a look this week.

**Comment by romgrk** [comment]
So it looks like we're using MUI Core's MenuList, which has an `autoFocus` prop. The problem is that it doesn't restore focus, and it also doesn't trap it inside itself. I'm not sure if it's intentionally missing or not. We'll need to add the logic either to that component, or to our internal component.

@mui/core Any thoughts on the above? Is there value in adding those capacities to the component?

**Comment by oliviertassinari** [comment]
This is intentional as far as I know. The root problem as far as I remember is the existence of a GridMenu component, this one shouldn't exist in the first place. It's related to https://github.com/mui/material-ui/issues/38756 and https://github.com/mui/material-ui/issues/38426#issuecomment-1699638504. We should likely have a brand new Popover primitive to replace GridMenu.

**Comment by romgrk** [comment]
We've integrated #10412 because that workaround works for now and solves the a11y issue. We should definitely replace `GridMenu` once the core has a primitive for that.

**Comment by Isaac1606** [comment]
https://github.com/mui/mui-x/pull/10412 The workaround does not work when using anchor tag, see https://github.com/mui/mui-x/issues/9913. The behavior with the current code is everything works fine when clicking on it with the mouse, but if you press enter nothing happens and the menuItem remains on the screen 
![image](https://github.com/mui/mui-x/assets/51690770/d437b9cc-2e57-431e-8d09-3442cce880f4)

**Comment by romgrk** [comment]
Do you have a live example for that?

**Comment by Isaac1606** [comment]
No, I do not have a live example, but it is pretty straightforward in a DataGrid define a GridActionsCellItem as it follows:
`<GridActionsCellItem showInMenu {...{component: "a", href: elem.href, target: "_blank"}} />`
And if you click on it works fine but if you use keyboard only the cursor goes back to the action's button and the menu keeps floating like in the Image that I previously attached.

**Comment by romgrk** [comment]
I don't think it's strictly related to the toolbar panels. If you want you can open a new issue for that. If you can link at least an example from the docs instead of a reproducible example, it also works.

**Comment by romgrk** [comment]
Let's track this in #10675, I think we need to discuss other keyboard interactions for GridActionsCellItem. You can subscribe to that issue to get notifications.

**Comment by HolyNoodle** [comment]
I have been struggling with a similar problem. I want the datagrid column menu to close when the user clicks on a menu item. This click triggers a focus in an input elsewhere in the page. 

Menu keeps stealing the focus. Even with event propagation stopped on multiple key board and mouse events. Couldn't pin the events I should bypass.

So I went for a different way: waiting for the event to finish bubbling by adding the code I want to trigger at the end of the call stack

```js
Promise.resolve().then(() => trigger focus action);
```
